narwhal/primary/src/tests/certificate_waiter_tests.rs

```diff
@@ -1,177 +1,90 @@
-    let (tx_parents, _rx_parents) = test_utils::test_channel!(100);
-
-    // Create test stores.
-    let (header_store, certificates_store, payload_store) = create_db_stores();
-
-    // Signal consensus round
-    let (_tx_consensus_round_updates, rx_consensus_round_updates) = watch::channel(0u64);
-
-    // Make a synchronizer for the core.
-    let synchronizer = Synchronizer::new(
-        name.clone(),
-        &committee,
-        certificates_store.clone(),
-        payload_store.clone(),
-        /* tx_header_waiter */ tx_sync_headers,
-        /* tx_certificate_waiter */ tx_sync_certificates,
-        None,
-    );
-
-    let network = anemo::Network::bind("localhost:0")
-        .server_name("narwhal")
-        .private_key(network_key)
-        .start(anemo::Router::new())
-        .unwrap();
-
-    let metrics = Arc::new(PrimaryMetrics::new(&Registry::new()));
-    let gc_depth: Round = 50;
-
-    // Make a headerWaiter
-    let _header_waiter_handle = HeaderWaiter::spawn(
-        name.clone(),
-        committee.clone(),
-        worker_cache.clone(),
-        certificates_store.clone(),
-        payload_store.clone(),
-        rx_consensus_round_updates.clone(),
-        gc_depth,
-        /* sync_retry_delay */ Duration::from_secs(5),
-        /* sync_retry_nodes */ 3,
-        rx_reconfigure.clone(),
-        rx_sync_headers,
-        tx_headers_loopback,
-        metrics.clone(),
-        P2pNetwork::new(network.clone()),
-    );
-
-    // Make a certificate waiter
-    let _certificate_waiter_handle = CertificateWaiter::spawn(
-        committee.clone(),
-        certificates_store.clone(),
-        rx_consensus_round_updates.clone(),
-        gc_depth,
-        rx_reconfigure.clone(),
-        rx_sync_certificates,
-        tx_certificates_loopback,
-        metrics.clone(),
-    );
+use tokio::{
+    sync::{
+        mpsc::{self, error::TryRecvError, Receiver, Sender},
+        watch, Mutex,
+    },
+    time::sleep,
+};
+use types::{
+    Certificate, FetchCertificatesRequest, FetchCertificatesResponse, PrimaryMessage,
+    ReconfigureNotification, Round,
+};
 
-    // Spawn the core.
-    let _core_handle = Core::spawn(
-        name.clone(),
-        committee.clone(),
-        worker_cache,
-        header_store.clone(),
-        certificates_store.clone(),
-        create_test_vote_store(),
-        synchronizer,
-        signature_service,
-        rx_consensus_round_updates,
-        /* gc_depth */ gc_depth,
-        rx_reconfigure,
-        /* rx_primaries */ rx_primary_messages,
-        /* rx_header_waiter */ rx_headers_loopback,
-        /* rx_certificate_waiter */ rx_certificates_loopback,
-        /* rx_proposer */ rx_headers,
-        tx_consensus,
-        /* tx_proposer */ tx_parents,
-        metrics.clone(),
-        P2pNetwork::new(network),
-    );
+struct FetchCertificateProxy {
```

Comment:
Following up on other comment, I think this is what should be replaceable with MockPrimaryToPrimary

Reply:
Sounds good, working on refactoring the test.